File: .github/workflows/validate-workflow-run-steps.js
// const parser = require('github-actions-parser');
const yaml = require('js-yaml');
const child_process = require('child_process');
const fs = require('fs/promises');
const path = require('path');

function checkBashSyntax (script) {
    return new Promise((resolve, _reject) => {
        const child = child_process.execFile(
            'bash',
            ['-c', '-n', script],
            (_error, stdout, stderr) => {
                resolve({
                    exitCode: child.exitCode,
                    stdout,
                    stderr
                });
            }
        );
    });
}

async function checkWorkflow (filepath) {
    const filename = path.basename(filepath);
    let data = null;
    try {
        const rawFile = await fs.readFile(filepath, 'utf8');
        data = yaml.load(rawFile);
    }
    catch (error) {
        console.error(`✘ Could not parse ${filepath} as YAML`);
        return 1;
    }

    let code = 0;
    for (const [name, job] of Object.entries(data.jobs || {})) {
        const steps = job.steps || [];
        for (let i = 0; i < steps.length; i++) {
            const step = steps[i];
            if (!step.run) continue;

            const stepName = `${filename} job "${name}", step "${step.name || i}"`
            const shell = step.shell || 'bash';
            if (!shell.startsWith('bash')) {
                console.error(`? ${stepName}`);
                console.error(`  Cannot evaluate non-bash run steps (shell: "${shell}")`);
                continue;
            }

            const result = await checkBashSyntax(step.run);
            if (result.exitCode === 0) {
                console.error(`✔︎ ${stepName}`);
            }
            else {
                code = 1
                console.error(`✘ ERROR: ${stepName}`);
                for (let line of result.stderr.split('\n')) {
                    // Lines are prefixed with 'bash: -c: '
                    line = line.slice('bash: -c: '.length);
                    console.error(`  ${line}`);
                }
            }
        }
    }

    return code;
}

async function checkDirectory (directory) {
    console.error(`Validating workflows in ${directory}...`);
    let code = 0;
    const files = await fs.readdir(directory);
    for (const filename of files) {
        if (['.yaml', '.yml'].includes(path.extname(filename))) {
            const newCode = await checkWorkflow(path.join(directory, filename));
            code = code || newCode;
        }
    }
    return code;
}


const directory = process.argv[2] || '.github/workflows';
checkDirectory(directory)
    .then(code => process.exit(code))
    .catch(error => {
        console.error(error.toString());
        process.exit(1);
    });

File: .github/workflows/validate-workflow-run-steps.py
"Check syntax of workflow run steps using bash in all workflows."
from pathlib import Path
import re
from subprocess import run
import sys
import yaml
from yaml import CLoader as Loader

def check_workflow(filepath):
    with filepath.open() as file:
        data = yaml.load(file, Loader=yaml.CLoader)

    exit_code = 0
    for name, job in data.get('jobs', {}).items():
        for index, step in enumerate(job.get('steps', [])):
            if 'run' not in step:
                continue
            
            step_name = f'{filepath.name} job "{name}", step "{step.get("name", index)}"'
            if not step.get('shell', 'bash').startswith('bash'):
                print(f'? {step_name}')
                print(f'  Cannot evaluate non-bash run steps (shell: "{step["shell"]}")')
                continue
            
            result = run(['bash', '-c', '-n', step['run']], capture_output=True)
            if result.returncode == 0:
                print(f'✔︎ {step_name}')
            if result.returncode > 0:
                exit_code = 1
                print(f'✘ ERROR: {step_name}')
                for line in result.stderr.decode().split('\n'):
                    line = line.partition('-c: ')[2] or line
                    print(f'  {line}')
    
    return exit_code

exit_code = 0
for child in Path('./.github/workflows/').iterdir():
    if child.suffix == '.yaml' or child.suffix == '.yml':
        exit_code = check_workflow(child) or exit_code

sys.exit(exit_code)
